File: 97.Pollard’sRhoFactorizationAlgo.py
# Pollard’s Rho is a probabilistic algorithm used to find a non-trivial factor of a
# composite number efficiently.
# It works using a pseudo-random function and Floyd’s cycle
# detection to identify repeating values that reveal factors.

# Explanation

# The algorithm relies on the idea that, for a composite number n, if two numbers x and y become congruent modulo a factor of n, we can find that factor using the greatest common divisor (GCD).

# We generate a sequence using a function like:

# f(x) = (x² + c) % n

# and track two sequences — one moves one step at a time (x),
# and the other moves two steps (y = f(f(y))).
# If they meet, we compute gcd(|x - y|, n) — this often reveals a factor of n.

import random

def gcd(a, b):
    if b == 0:
        return a
    return gcd(b, a % b)


def pollard_rho(n):
    if n % 2 == 0:
        return 2

    c = random.randint(1, n - 1)

    x = random.randint(0, n - 1)
    y = x

    d = 1

    while d == 1:
        x = (x * x + c) % n
        y = (y * y + c) % n
        y = (y * y + c) % n

        d = gcd(abs(x - y), n)

        if d == n:
            return pollard_rho(n)

    return d


n = 8051
factor = pollard_rho(n)
print("A non-trivial factor of", n, "is:", factor)

File: 97.Pollard’sRhoFactorizationAlgo.java
/* ----------------------------------------------------------------------------
   ( The Authentic JS/JAVA CodeBuff )
    ___ _                      _              _ 
    | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
    | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
    |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                           |__/ 
   ----------------------------------------------------------------------------
   Youtube: https://youtube.com/@code-with-Bharadwaj
   Github : https://github.com/Manu577228
   Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio
   -------------------------------------------------------------------------- */

import java.io.*;
import java.util.*;

public class Main {

    // Function to compute GCD using Euclidean Algorithm
    static long gcd(long a, long b) {
        if (b == 0) return a;
        return gcd(b, a % b);
    }

    // Pollard's Rho Algorithm
    static long pollardRho(long n) {
        // If number is even, return 2 as factor
        if (n % 2 == 0)
            return 2;

        Random rand = new Random();
        long c = rand.nextInt((int) (n - 1)) + 1; // Choose random c (1 < c < n)
        long x = rand.nextInt((int) (n - 1));     // Random initial x
        long y = x;
        long d = 1;

        // f(x) = (x*x + c) % n repeated until factor found
        while (d == 1) {
            x = (mulmod(x, x, n) + c) % n;        // Move x one step
            y = (mulmod(y, y, n) + c) % n;        // Move y two steps (Floyd's cycle)
            y = (mulmod(y, y, n) + c) % n;

            d = gcd(Math.abs(x - y), n);          // Compute gcd(|x - y|, n)

            if (d == n)                           // Retry if failed
                return pollardRho(n);
        }

        return d;  // Found non-trivial factor
    }

    // Function to perform modular multiplication safely
    static long mulmod(long a, long b, long mod) {
        long res = 0;
        a %= mod;
        while (b > 0) {
            if ((b & 1) == 1)
                res = (res + a) % mod;
            a = (2 * a) % mod;
            b >>= 1;
        }
        return res % mod;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);

        long n = 8051; // Example number (8051 = 97 * 83)
        long factor = pollardRho(n);

        out.println("A non-trivial factor of " + n + " is: " + factor);
        out.flush();
    }
}
